{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Hard Coded Domain Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# make df of sars-cov-2 orf1ab protein products and their start and end aa positions within orf1ab\n",
    "# positions taken from Wu et al. 2022: https://doi.org/10.1038/s41401-021-00851-w\n",
    "\n",
    "proteins = [ \"NSP1\", \"NSP2\", \"NSP3\", \"NSP4\", \"NSP5\", \"NSP6\", \"NSP7\", \"NSP8\", \"NSP9\", \"NSP10\", \"NSP12\", \"NSP13\", \"NSP14\", \"NSP15\", \"NSP16\"]\n",
    "\n",
    "proteinStarts = [1, 181, 819, 2764, 3264, 3570, 3860, 3943, 4141, 4254, 4393, 5325, 5926, 6453, 6799]\n",
    "\n",
    "proteinEnds = [180, 818, 2763, 3263, 3569, 3859, 3942, 4140, 4253, 4392, 5324, 5925, 6452, 6798, 7096]\n",
    "\n",
    "proteinsData = {'Protein':proteins, 'Protein Starts':proteinStarts, 'Protein Ends':proteinEnds}\n",
    "\n",
    "proteinsdf = pd.DataFrame(data=proteinsData)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# spike\n",
    "spikeDomains = [ \"s-no-domain\", \"NTD\", \"s-no-domain\", \"RBD\", \"s-no-domain\", \"FP\", \"s-no-domain\", \"HR1\", \"s-no-domain\", \"HR2\", \"TMD\", \"CD\"]\n",
    "spikeDomainStarts = [1, 13, 306, 319, 542, 788, 807, 912, 985, 1163, 1213, 1237]\n",
    "spikeDomainEnds = [12, 305, 318, 541, 787, 806, 911, 984, 1162, 1212, 1236, 1273]\n",
    "\n",
    "# Make data frame from those lists and give them appropriate column names\n",
    "spikeDomainData = {'Domains':spikeDomains, 'Domain Starts':spikeDomainStarts, 'Domain Ends':spikeDomainEnds}\n",
    "spikeDomainDF = pd.DataFrame(data=spikeDomainData)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# envelope - https://www.nature.com/articles/s41594-020-00536-8/figures/1\n",
    "eDomains = [\"NTD\", \"TMD\", \"CTD\"]\n",
    "eDomainStarts = [1, 8, 39]\n",
    "eDomainEnds = [7, 38, 75]\n",
    "\n",
    "# Make data frame from those lists and give them appropriate column names\n",
    "eDomainData = {'Domains':eDomains, 'Domain Starts':eDomainStarts, 'Domain Ends':eDomainEnds}\n",
    "eDomainDF = pd.DataFrame(data=eDomainData)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# nucleocapsid - https://www.nature.com/articles/s41467-021-21953-3#Fig1\n",
    "nDomains = [\"NTD\", \"RBD\", \"LINK\", \"Dimerisation\", \"CTD\"]\n",
    "nDomainStarts = [1, 50, 174, 247, 365 ]\n",
    "nDomainEnds = [49, 173, 246, 364, 419]\n",
    "\n",
    "# Make data frame from those lists and give them appropriate column names\n",
    "nDomainData = {'Domains':nDomains, 'Domain Starts':nDomainStarts, 'Domain Ends':nDomainEnds}\n",
    "nDomainDF = pd.DataFrame(data=nDomainData)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# membrane - https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8221690/\n",
    "mDomains = [\"n-terminal ecto\", \"TMI\", \"TMI-TMII Link\", \"TMII\",\"TMII-TMIII Link\", \"TMIII\", \"m-no-domain\", \"cytoplasmic\"]\n",
    "mDomainStarts = [1, 20, 40, 51, 74, 78, 101, 118]\n",
    "mDomainEnds = [19, 38, 50, 73, 77, 100, 117, 222]\n",
    "\n",
    "# Make data frame from those lists and give them appropriate column names\n",
    "mDomainData = {'Domains':mDomains, 'Domain Starts':mDomainStarts, 'Domain Ends':mDomainEnds}\n",
    "mDomainDF = pd.DataFrame(data=mDomainData)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# orf3a: https://doi.org/10.3389/fmicb.2022.854567, https://doi.org/10.1128/mSystems.00266-20\n",
    "\n",
    "orf3aSubDoms = [ 'orf3a-NTD', 'orf3a-TMDI-III', 'orf3a-no-domain', 'orf3a-β1–β8', 'orf3a-no-domain', 'orf3a-CTD']\n",
    "orf3aSubDomsStart = [ 1, 40, 129, 145, 236, 239]\n",
    "orf3aSubDomsEnd = [39, 128, 144, 235, 238, 276]\n",
    "\n",
    "orf3aSubDomData = {'Domains':orf3aSubDoms, 'Domain Starts':orf3aSubDomsStart, 'Domain Ends':orf3aSubDomsEnd}\n",
    "orf3aSubDomsDF = pd.DataFrame(data= orf3aSubDomData)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# orf7a: https://www.cell.com/iscience/pdf/S2589-0042(21)00155-3.pdf\n",
    "\n",
    "orf7aSubDoms = ['orf7a-N-terminal', 'orf7a-Ig-like-ectodomain', 'orf7a-hydrophobic-TMD', 'orf7a-ER-retention-motif']\n",
    "orf7aSubDomsStart = [1, 16, 97, 117]\n",
    "orf7aSubDomsEnd = [15, 96, 116, 122]\n",
    "\n",
    "orf7aSubDomData = {'Domains': orf7aSubDoms, 'Domain Starts': orf7aSubDomsStart, 'Domain Ends': orf7aSubDomsEnd}\n",
    "orf7aSubDomsDF = pd.DataFrame(data=orf7aSubDomData)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# orf8\n",
    "orf8SubDoms = ['orf8-N-terminal-sig-seq', 'orf8-Ig-like-domain']\n",
    "orf8SubDomsStart = [1, 16]\n",
    "orf8SubDomsEnd = [15, 122]\n",
    "\n",
    "orf8SubDomData = {'Domains':orf8SubDoms, 'Domain Starts':orf8SubDomsStart, 'Domain Ends':orf8SubDomsEnd}\n",
    "orf8SubDomsDF = pd.DataFrame(data= orf8SubDomData)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# non-structural protein domains (NSPs)\n",
    "# only nsp1 will start from 1, need to add the aa position of the nsp within orf1ab to each of the others\n",
    "# eg position 1 in nsp3 will be 819\n",
    "\n",
    "nspSubDoms = [ 'NSP1', 'NSP2', 'NSP3-Ubl1', 'NSP3-HVR','NSP3-MacI', 'NSP3-no-domain', 'NSP3-MacII', 'NSP3-MacIII', 'NSP3-DPUP', 'NSP3-Ubl2', 'NSP3-PL2pro', 'NSP3-no-domain', 'NSP3-NAB', 'NSP3-Beta-SM', 'NSP3-TM', 'NSP3-no-domain', 'NSP3-Subdomain-Y', 'NSP4-no-domain', 'NSP4-TM', 'NSP4-no-domain', 'NSP5-no-domain', 'NSP5-Mpro', 'NSP5-no-domain', 'NSP6-TM', 'NSP7', 'NSP8', 'NSP9', 'NSP10', 'NSP12-no-domain', 'NSP12-NiRAN', 'NSP12-Interface', 'NSP12-RdRP', 'NSP13-ZBD', 'NSP13-no-domain', 'NSP13-Helicase', 'NSP13-no-domain', 'NSP14', 'NSP15-no-domain', 'NSP15-EndoU', 'NSP15-no-domain', 'NSP16' ]\n",
    "nspSubDomsStart = [1, 181, 819, 930, 1025, 1199, 1232, 1369, 1494, 1564, 1623, 1882, 1907, 2022, 2231, 2373, 2402, 2764, 2777, 3159, 3264, 3267, 3564, 3570, 3860, 3943, 4141, 4254, 4393, 4444, 4643, 4759, 5325, 5420, 5675, 5915, 5926, 6453, 6646, 6797, 6799]\n",
    "nspSubDomsEnd = [180, 818, 929, 1024, 1198, 1231, 1368, 1493, 1563, 1622, 1881, 1906, 2021, 2230, 2372, 2401, 2763, 2776, 3158, 3263, 3266, 3563, 3569, 3859, 3942, 4140, 4253, 4392, 4443, 4642, 4758, 5324, 5419, 5674, 5914, 5925, 6452, 6645, 6796, 6798, 7096]\n",
    "\n",
    "nspSubDomnData = {'Domains':nspSubDoms, 'Domain Starts':nspSubDomsStart, 'Domain Ends':nspSubDomsEnd}\n",
    "nspSubDomnDF = pd.DataFrame(data=nspSubDomnData)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "File Parsing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'pd' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[1], line 26\u001b[0m\n\u001b[1;32m     24\u001b[0m \u001b[38;5;66;03m# call function\u001b[39;00m\n\u001b[1;32m     25\u001b[0m mutations_file \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m/home4/2191618l/Github/Chapter3/CO_MU_SARS-CoV-2/Output/Generated_Sequences/Reference/mutations.csv\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[0;32m---> 26\u001b[0m df \u001b[38;5;241m=\u001b[39m \u001b[43mprocess_mutations_csv\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmutations_file\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     27\u001b[0m \u001b[38;5;66;03m#Remove recombinants\u001b[39;00m\n\u001b[1;32m     28\u001b[0m df \u001b[38;5;241m=\u001b[39m df[df\u001b[38;5;241m.\u001b[39mSequence_Name\u001b[38;5;241m.\u001b[39mstr\u001b[38;5;241m.\u001b[39mcontains(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mX\u001b[39m\u001b[38;5;124m'\u001b[39m) \u001b[38;5;241m==\u001b[39m \u001b[38;5;28;01mFalse\u001b[39;00m]\n",
      "Cell \u001b[0;32mIn[1], line 4\u001b[0m, in \u001b[0;36mprocess_mutations_csv\u001b[0;34m(mut_file)\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mprocess_mutations_csv\u001b[39m(mut_file):\n\u001b[1;32m      3\u001b[0m     \u001b[38;5;66;03m# Read in csv file containing mutations\u001b[39;00m\n\u001b[0;32m----> 4\u001b[0m     df \u001b[38;5;241m=\u001b[39m \u001b[43mpd\u001b[49m\u001b[38;5;241m.\u001b[39mread_csv(mut_file)\n\u001b[1;32m      6\u001b[0m     \u001b[38;5;66;03m# Get rid of duplicated sequence name information from the Sequence_Name column\u001b[39;00m\n\u001b[1;32m      7\u001b[0m     df[[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mSequence_Name\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124ms1\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124ms2\u001b[39m\u001b[38;5;124m'\u001b[39m]] \u001b[38;5;241m=\u001b[39m df[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mSequence_Name\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m.\u001b[39mstr\u001b[38;5;241m.\u001b[39msplit(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m|\u001b[39m\u001b[38;5;124m\"\u001b[39m, expand\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'pd' is not defined"
     ]
    }
   ],
   "source": [
    "# parse mutations file and process it - fix sequence name, remove Nan amino acid valies, add two columns: reference and alternative aa\n",
    "def process_mutations_csv(mut_file):\n",
    "    # Read in csv file containing mutations\n",
    "    df = pd.read_csv(mut_file)\n",
    "    \n",
    "    # Get rid of duplicated sequence name information from the Sequence_Name column\n",
    "    df[['Sequence_Name', 's1', 's2']] = df['Sequence_Name'].str.split(\"|\", expand=True)\n",
    "    df = df.drop(['s1', 's2'], axis=1)\n",
    "    \n",
    "    # Removing columns where Amino_Acid value is NaN\n",
    "    df = df[~df['Amino_Acid'].isna()]\n",
    "    \n",
    "    # Add alt_aa position column\n",
    "    alt_aa_list = df[\"Amino_Acid\"].values.tolist()\n",
    "    alternativePosList = [x[-1:] for x in alt_aa_list]\n",
    "\n",
    "    # Add ref_aa position column\n",
    "    refPosList = [x.split(\":\")[1][0] for x in alt_aa_list]\n",
    "    df[\"REF_AA\"] = refPosList\n",
    "    df[\"ALT_AA\"] = alternativePosList\n",
    "        \n",
    "    return df\n",
    "\n",
    "# call function\n",
    "mutations_file = '/home4/2191618l/Github/Chapter3/CO_MU_SARS-CoV-2/Output/Generated_Sequences/Reference/mutations.csv'\n",
    "df = process_mutations_csv(mutations_file)\n",
    "#Remove recombinants\n",
    "df = df[df.Sequence_Name.str.contains('X') == False]\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to add protein column to mutations df\n",
    "def assign_protein(df, proteinsdf):\n",
    "    proteinList = []\n",
    "\n",
    "    # check which NSP the amino acid is within the protein positions for and append..\n",
    "    for index, row in df.iterrows():\n",
    "        if row['ORF'] == 'orf1ab':\n",
    "            for _, protein_row in proteinsdf.iterrows():\n",
    "                if row['AMINO_ORF_POS'] >= protein_row['Protein Starts'] and row['AMINO_ORF_POS'] <= protein_row['Protein Ends']:\n",
    "                    proteinList.append(protein_row['Protein'])\n",
    "                    break\n",
    "        else:\n",
    "            proteinList.append(row['ORF'])\n",
    "\n",
    "    df['Protein'] = proteinList\n",
    "\n",
    "# call function\n",
    "assign_protein(df, proteinsdf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to add sliding window column to mutations df\n",
    "\n",
    "# Need to give the name of the ORF and the dataframe that has its mutations\n",
    "def SlidingWindow(df, WindowLength):\n",
    "    # Store the calculated windows in a list\n",
    "    windows = []\n",
    "    # loop through each row of the mutation dataframe, to calculate window of each mutation\n",
    "    for index, row in df.iterrows():\n",
    "        window = (int)(row['AMINO_ORF_POS'] / WindowLength) + 1\n",
    "        # add calculated window to list of calculated windows\n",
    "        windows.append(window)\n",
    "\n",
    "    # add calculated windows as new column 'Window' in current ORFs mutation dataframe\n",
    "    df['Window'] = windows\n",
    "\n",
    "SlidingWindow(df, 20)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Adding sub-domain information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [],
   "source": [
    "# make a dataframe for each ORF containing all its mutations\n",
    "# append each of the positions within an orf to a list \n",
    "\n",
    "# ORF1ab\n",
    "orf1abdf = (df.loc[(df['ORF'] == 'orf1ab')])\n",
    "orf1abdfList = list(orf1abdf.AMINO_ORF_POS)\n",
    "\n",
    "# S\n",
    "sdf = (df.loc[(df['ORF'] == 'S')])\n",
    "sdfList = list(sdf.AMINO_ORF_POS)\n",
    "\n",
    "# N\n",
    "ndf = (df.loc[(df['ORF'] == 'N')])\n",
    "ndfList = list(ndf.AMINO_ORF_POS)\n",
    "\n",
    "# E\n",
    "edf = (df.loc[(df['ORF'] == 'E')])\n",
    "edfList = list(edf.AMINO_ORF_POS)\n",
    "\n",
    "# M\n",
    "mdf = (df.loc[(df['ORF'] == 'M')])\n",
    "mdfList = list(mdf.AMINO_ORF_POS)\n",
    "\n",
    "# ORF3a\n",
    "orf3adf = (df.loc[(df['ORF'] == 'ORF3a')])\n",
    "orf3aList = list(orf3adf.AMINO_ORF_POS)\n",
    "\n",
    "# orf6\n",
    "orf6df = (df.loc[(df['ORF'] == 'ORF6')])\n",
    "orf6List = list(orf6df.AMINO_ORF_POS)\n",
    "\n",
    "# orf7a\n",
    "orf7adf = (df.loc[(df['ORF'] == 'ORF7a')])\n",
    "orf7aList = list(orf7adf.AMINO_ORF_POS)\n",
    "\n",
    "# orf8\n",
    "orf8df = (df.loc[(df['ORF'] == 'ORF8')])\n",
    "orf8List = list(orf8df.AMINO_ORF_POS)\n",
    "\n",
    "# orf10\n",
    "orf10df = (df.loc[(df['ORF'] == 'ORF10')])\n",
    "orf10List = list(orf10df.AMINO_ORF_POS)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "# function to append domains for each mutation\n",
    "def sub_dom_function(orfList, domainDF, orfDF):\n",
    "    domList = []\n",
    "    for mut in orfList:\n",
    "        for index, row in domainDF.iterrows():\n",
    "            if mut >= row['Domain Starts'] and mut <= row['Domain Ends']:\n",
    "                domList.append(row['Domains'])\n",
    "                break\n",
    "\n",
    "    orfDF['Sub-Domains'] = domList"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_2665646/2271704794.py:10: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  orfDF['Sub-Domains'] = domList\n",
      "/tmp/ipykernel_2665646/2271704794.py:10: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  orfDF['Sub-Domains'] = domList\n",
      "/tmp/ipykernel_2665646/2271704794.py:10: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  orfDF['Sub-Domains'] = domList\n",
      "/tmp/ipykernel_2665646/2271704794.py:10: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  orfDF['Sub-Domains'] = domList\n",
      "/tmp/ipykernel_2665646/2271704794.py:10: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  orfDF['Sub-Domains'] = domList\n",
      "/tmp/ipykernel_2665646/2271704794.py:10: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  orfDF['Sub-Domains'] = domList\n",
      "/tmp/ipykernel_2665646/2271704794.py:10: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  orfDF['Sub-Domains'] = domList\n",
      "/tmp/ipykernel_2665646/2271704794.py:10: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  orfDF['Sub-Domains'] = domList\n"
     ]
    }
   ],
   "source": [
    "sub_dom_function(sdfList, spikeDomainDF, sdf)\n",
    "sub_dom_function(edfList, eDomainDF, edf)\n",
    "sub_dom_function(ndfList, nDomainDF, ndf)\n",
    "sub_dom_function(mdfList, mDomainDF, mdf)\n",
    "sub_dom_function(orf3aList, orf3aSubDomsDF, orf3adf)\n",
    "sub_dom_function(orf8List, orf8SubDomsDF, orf8df)\n",
    "sub_dom_function(orf7aList, orf7aSubDomsDF, orf7adf)\n",
    "sub_dom_function(orf1abdfList, nspSubDomnDF, orf1abdf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_2665646/2225669422.py:11: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  orf6df['Sub-Domains'] = orf6SubDomList\n",
      "/tmp/ipykernel_2665646/2225669422.py:15: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  orf10df['Sub-Domains'] = orf10SubDomList\n"
     ]
    }
   ],
   "source": [
    "# orf6 and orf10\n",
    "# function to append sub-domain for ORF, where there are no sub-domains\n",
    "def subdomain_list(orf, n):\n",
    "    subdomainList = []\n",
    "    for i in orf:\n",
    "        subdomainList.append(n)\n",
    "    return subdomainList\n",
    "\n",
    "# adding sub domains for orf6 to dataframe\n",
    "orf6SubDomList = subdomain_list(orf6List, 'orf6')\n",
    "orf6df['Sub-Domains'] = orf6SubDomList\n",
    "\n",
    "# adding sub domains for orf10 to dataframe\n",
    "orf10SubDomList = subdomain_list(orf10List, 'orf10')\n",
    "orf10df['Sub-Domains'] = orf10SubDomList"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "# concatenate the dfs\n",
    "\n",
    "# final dataframe containing each mutations with annotations for sliding window, sub-domain and protein\n",
    "concatenatedDF = pd.concat([orf1abdf, orf3adf, orf6df, orf7adf, orf8df, orf10df, sdf, edf, ndf, mdf], axis=0)\n",
    "# convert mutation position to an integer\n",
    "concatenatedDF['AMINO_ORF_POS'] = df['AMINO_ORF_POS'].astype('int')\n",
    "# group the dataframe by sequence name \n",
    "concatenatedDF['pos'] = concatenatedDF.groupby('Sequence_Name').ngroup()\n",
    "concatenatedDF = concatenatedDF.sort_values(by='pos').reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence_Name</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>POS</th>\n",
       "      <th>BIO_POS</th>\n",
       "      <th>AMINO_ORF_POS</th>\n",
       "      <th>Nucleotide</th>\n",
       "      <th>Nucleotide_Context</th>\n",
       "      <th>Amino_Acid</th>\n",
       "      <th>Amino_Acid_Context</th>\n",
       "      <th>...</th>\n",
       "      <th>Codon</th>\n",
       "      <th>ORF</th>\n",
       "      <th>Synonymous</th>\n",
       "      <th>Ambiguous</th>\n",
       "      <th>REF_AA</th>\n",
       "      <th>ALT_AA</th>\n",
       "      <th>Window</th>\n",
       "      <th>Protein</th>\n",
       "      <th>Sub-Domains</th>\n",
       "      <th>pos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>28143</td>\n",
       "      <td>28144</td>\n",
       "      <td>84</td>\n",
       "      <td>T28144C</td>\n",
       "      <td>TC-TTA</td>\n",
       "      <td>ORF8:L84S</td>\n",
       "      <td>LS-CLP</td>\n",
       "      <td>...</td>\n",
       "      <td>tTa</td>\n",
       "      <td>ORF8</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>L</td>\n",
       "      <td>S</td>\n",
       "      <td>5</td>\n",
       "      <td>ORF8</td>\n",
       "      <td>orf8-Ig-like-domain</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A.1</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>17857</td>\n",
       "      <td>17858</td>\n",
       "      <td>1464</td>\n",
       "      <td>A17858G</td>\n",
       "      <td>AG-TAT</td>\n",
       "      <td>orf1ab:Y1464C</td>\n",
       "      <td>YC-EYD</td>\n",
       "      <td>...</td>\n",
       "      <td>tAt</td>\n",
       "      <td>orf1ab</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>Y</td>\n",
       "      <td>C</td>\n",
       "      <td>74</td>\n",
       "      <td>NSP3</td>\n",
       "      <td>NSP3-MacIII</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A.1</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>17746</td>\n",
       "      <td>17747</td>\n",
       "      <td>1427</td>\n",
       "      <td>C17747T</td>\n",
       "      <td>CT-CCT</td>\n",
       "      <td>orf1ab:P1427L</td>\n",
       "      <td>PL-NPA</td>\n",
       "      <td>...</td>\n",
       "      <td>cCt</td>\n",
       "      <td>orf1ab</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>P</td>\n",
       "      <td>L</td>\n",
       "      <td>72</td>\n",
       "      <td>NSP3</td>\n",
       "      <td>NSP3-MacIII</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A.11</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>28877</td>\n",
       "      <td>28878</td>\n",
       "      <td>202</td>\n",
       "      <td>G28878A</td>\n",
       "      <td>GA-AGT</td>\n",
       "      <td>N:S202N</td>\n",
       "      <td>SN-SSR</td>\n",
       "      <td>...</td>\n",
       "      <td>aGt</td>\n",
       "      <td>N</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>S</td>\n",
       "      <td>N</td>\n",
       "      <td>11</td>\n",
       "      <td>N</td>\n",
       "      <td>LINK</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A.12</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>26427</td>\n",
       "      <td>26428</td>\n",
       "      <td>62</td>\n",
       "      <td>G26428T</td>\n",
       "      <td>GT-TGT</td>\n",
       "      <td>E:V62F</td>\n",
       "      <td>VF-RVK</td>\n",
       "      <td>...</td>\n",
       "      <td>Gtt</td>\n",
       "      <td>E</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>V</td>\n",
       "      <td>F</td>\n",
       "      <td>4</td>\n",
       "      <td>E</td>\n",
       "      <td>CTD</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9229</th>\n",
       "      <td>W.4</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>20598</td>\n",
       "      <td>20599</td>\n",
       "      <td>2378</td>\n",
       "      <td>A20599C</td>\n",
       "      <td>AC-AAT</td>\n",
       "      <td>orf1ab:I2378L</td>\n",
       "      <td>IL-EIS</td>\n",
       "      <td>...</td>\n",
       "      <td>Att</td>\n",
       "      <td>orf1ab</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>I</td>\n",
       "      <td>L</td>\n",
       "      <td>119</td>\n",
       "      <td>NSP3</td>\n",
       "      <td>NSP3-no-domain</td>\n",
       "      <td>1861</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9230</th>\n",
       "      <td>W.4</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>28370</td>\n",
       "      <td>28371</td>\n",
       "      <td>33</td>\n",
       "      <td>G28371T</td>\n",
       "      <td>GT-AGT</td>\n",
       "      <td>N:S33I</td>\n",
       "      <td>SI-RSG</td>\n",
       "      <td>...</td>\n",
       "      <td>aGt</td>\n",
       "      <td>N</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>S</td>\n",
       "      <td>I</td>\n",
       "      <td>2</td>\n",
       "      <td>N</td>\n",
       "      <td>NTD</td>\n",
       "      <td>1861</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9231</th>\n",
       "      <td>Y.1</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>5258</td>\n",
       "      <td>5259</td>\n",
       "      <td>1665</td>\n",
       "      <td>C5259T</td>\n",
       "      <td>CT-ACT</td>\n",
       "      <td>orf1ab:T1665I</td>\n",
       "      <td>TI-LTS</td>\n",
       "      <td>...</td>\n",
       "      <td>aCt</td>\n",
       "      <td>orf1ab</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>T</td>\n",
       "      <td>I</td>\n",
       "      <td>84</td>\n",
       "      <td>NSP3</td>\n",
       "      <td>NSP3-PL2pro</td>\n",
       "      <td>1862</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9232</th>\n",
       "      <td>Z.1</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>25888</td>\n",
       "      <td>25889</td>\n",
       "      <td>166</td>\n",
       "      <td>C25889T</td>\n",
       "      <td>CT-TCA</td>\n",
       "      <td>ORF3a:S166L</td>\n",
       "      <td>SL-SSI</td>\n",
       "      <td>...</td>\n",
       "      <td>tCa</td>\n",
       "      <td>ORF3a</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>S</td>\n",
       "      <td>L</td>\n",
       "      <td>9</td>\n",
       "      <td>ORF3a</td>\n",
       "      <td>orf3a-β1–β8</td>\n",
       "      <td>1863</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9233</th>\n",
       "      <td>Z.1</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>25646</td>\n",
       "      <td>25647</td>\n",
       "      <td>85</td>\n",
       "      <td>G25647C</td>\n",
       "      <td>GC-TGT</td>\n",
       "      <td>ORF3a:L85F</td>\n",
       "      <td>LF-LLL</td>\n",
       "      <td>...</td>\n",
       "      <td>ttG</td>\n",
       "      <td>ORF3a</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>L</td>\n",
       "      <td>F</td>\n",
       "      <td>5</td>\n",
       "      <td>ORF3a</td>\n",
       "      <td>orf3a-TMDI-III</td>\n",
       "      <td>1863</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9234 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Sequence_Name REF ALT    POS  BIO_POS  AMINO_ORF_POS Nucleotide  \\\n",
       "0                A   T   C  28143    28144             84    T28144C   \n",
       "1              A.1   A   G  17857    17858           1464    A17858G   \n",
       "2              A.1   C   T  17746    17747           1427    C17747T   \n",
       "3             A.11   G   A  28877    28878            202    G28878A   \n",
       "4             A.12   G   T  26427    26428             62    G26428T   \n",
       "...            ...  ..  ..    ...      ...            ...        ...   \n",
       "9229           W.4   A   C  20598    20599           2378    A20599C   \n",
       "9230           W.4   G   T  28370    28371             33    G28371T   \n",
       "9231           Y.1   C   T   5258     5259           1665     C5259T   \n",
       "9232           Z.1   C   T  25888    25889            166    C25889T   \n",
       "9233           Z.1   G   C  25646    25647             85    G25647C   \n",
       "\n",
       "     Nucleotide_Context     Amino_Acid Amino_Acid_Context  ...  Codon     ORF  \\\n",
       "0                TC-TTA      ORF8:L84S             LS-CLP  ...    tTa    ORF8   \n",
       "1                AG-TAT  orf1ab:Y1464C             YC-EYD  ...    tAt  orf1ab   \n",
       "2                CT-CCT  orf1ab:P1427L             PL-NPA  ...    cCt  orf1ab   \n",
       "3                GA-AGT        N:S202N             SN-SSR  ...    aGt       N   \n",
       "4                GT-TGT         E:V62F             VF-RVK  ...    Gtt       E   \n",
       "...                 ...            ...                ...  ...    ...     ...   \n",
       "9229             AC-AAT  orf1ab:I2378L             IL-EIS  ...    Att  orf1ab   \n",
       "9230             GT-AGT         N:S33I             SI-RSG  ...    aGt       N   \n",
       "9231             CT-ACT  orf1ab:T1665I             TI-LTS  ...    aCt  orf1ab   \n",
       "9232             CT-TCA    ORF3a:S166L             SL-SSI  ...    tCa   ORF3a   \n",
       "9233             GC-TGT     ORF3a:L85F             LF-LLL  ...    ttG   ORF3a   \n",
       "\n",
       "     Synonymous  Ambiguous  REF_AA ALT_AA Window  Protein  \\\n",
       "0         False      False       L      S      5     ORF8   \n",
       "1         False      False       Y      C     74     NSP3   \n",
       "2         False      False       P      L     72     NSP3   \n",
       "3         False      False       S      N     11        N   \n",
       "4         False      False       V      F      4        E   \n",
       "...         ...        ...     ...    ...    ...      ...   \n",
       "9229      False      False       I      L    119     NSP3   \n",
       "9230      False      False       S      I      2        N   \n",
       "9231      False      False       T      I     84     NSP3   \n",
       "9232      False      False       S      L      9    ORF3a   \n",
       "9233      False      False       L      F      5    ORF3a   \n",
       "\n",
       "              Sub-Domains   pos  \n",
       "0     orf8-Ig-like-domain     0  \n",
       "1             NSP3-MacIII     1  \n",
       "2             NSP3-MacIII     1  \n",
       "3                    LINK     2  \n",
       "4                     CTD     3  \n",
       "...                   ...   ...  \n",
       "9229       NSP3-no-domain  1861  \n",
       "9230                  NTD  1861  \n",
       "9231          NSP3-PL2pro  1862  \n",
       "9232          orf3a-β1–β8  1863  \n",
       "9233       orf3a-TMDI-III  1863  \n",
       "\n",
       "[9234 rows x 21 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "concatenatedDF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "# function to convert list of lists to list of strings, each element in the string separated by a pipe\n",
    "# this will be used to create piped annotations \n",
    "def list_of_strings(l):\n",
    "    str_l = [\"|\".join([str(j) for j in i]) for i in l]\n",
    "    return str_l"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Sequence_Name', 'REF', 'ALT', 'POS', 'BIO_POS', 'AMINO_ORF_POS',\n",
       "       'Nucleotide', 'Nucleotide_Context', 'Amino_Acid', 'Amino_Acid_Context',\n",
       "       'Codon_Position', 'Codon', 'ORF', 'Synonymous', 'Ambiguous', 'REF_AA',\n",
       "       'ALT_AA', 'Window', 'Protein', 'Sub-Domains', 'pos'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "concatenatedDF.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "# add column to the dataframe called annotations\n",
    "# will have pipe separated annotations in the correct order to be run through WESME tool\n",
    "annotation_order = ['Nucleotide','ALT_AA', 'AMINO_ORF_POS', 'REF_AA', 'Window', 'Sub-Domains', 'Protein', 'ORF']\n",
    "concatenatedDF[\"annotations\"] = list_of_strings([list(row[annotation_order]) for i,row in concatenatedDF.iterrows()])\n",
    "annotation_order = ['ALT_AA','AMINO_ORF_POS', 'REF_AA', 'Window', 'Sub-Domains', 'Protein', 'ORF']\n",
    "concatenatedDF[\"amino_annotations\"] = list_of_strings([list(row[annotation_order]) for i,row in concatenatedDF.iterrows()])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0        S|84|L|5|orf8-Ig-like-domain|ORF8|ORF8\n",
       "1           C|1464|Y|74|NSP3-MacIII|NSP3|orf1ab\n",
       "2           L|1427|P|72|NSP3-MacIII|NSP3|orf1ab\n",
       "3                           N|202|S|11|LINK|N|N\n",
       "4                              F|62|V|4|CTD|E|E\n",
       "                         ...                   \n",
       "9229    L|2378|I|119|NSP3-no-domain|NSP3|orf1ab\n",
       "9230                           I|33|S|2|NTD|N|N\n",
       "9231        I|1665|T|84|NSP3-PL2pro|NSP3|orf1ab\n",
       "9232          L|166|S|9|orf3a-β1–β8|ORF3a|ORF3a\n",
       "9233        F|85|L|5|orf3a-TMDI-III|ORF3a|ORF3a\n",
       "Name: amino_annotations, Length: 9234, dtype: object"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "concatenatedDF.amino_annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create file output in right format to be run through WESME\n",
    "fileContents = []\n",
    "\n",
    "# first create header line\n",
    "samples = list(concatenatedDF.Sequence_Name.unique())\n",
    "headerString = 'samples' + \" \" + ','.join(samples)\n",
    "fileContents.append(headerString)\n",
    "\n",
    "for mutation in concatenatedDF.annotations.unique():\n",
    "    filtereddf = concatenatedDF[concatenatedDF.annotations == mutation]\n",
    "    mut = str(filtereddf.annotations.iloc[0]).replace(\" \",\"-\") + \" \"\n",
    "    mutList = []\n",
    "    for lineage in filtereddf.Sequence_Name.unique(): # added .unique()\n",
    "        lineageindex = samples.index(lineage)\n",
    "        mutList.append(str(lineageindex))\n",
    "    mut = mut + (','.join(mutList))\n",
    "    fileContents.append(mut)\n",
    "\n",
    "# Create file output in right format to be run through WESME\n",
    "aminoFileContents = []\n",
    "\n",
    "# first create header line\n",
    "samples = list(concatenatedDF.Sequence_Name.unique())\n",
    "headerString = 'samples' + \" \" + ','.join(samples)\n",
    "aminoFileContents.append(headerString)\n",
    "\n",
    "for mutation in concatenatedDF.annotations.unique():\n",
    "    filtereddf = concatenatedDF[concatenatedDF.annotations == mutation]\n",
    "    mut = str(filtereddf.annotations.iloc[0]).replace(\" \",\"-\") + \" \"\n",
    "    mutList = []\n",
    "    for lineage in filtereddf.Sequence_Name.unique(): # added .unique()\n",
    "        lineageindex = samples.index(lineage)\n",
    "        mutList.append(str(lineageindex))\n",
    "    mut = mut + (','.join(mutList))\n",
    "    aminoFileContents.append(mut)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "# saving file contents list as a .txt file\n",
    "with open(\"SARS_REF_smut_list.txt\", mode=\"w\") as file:\n",
    "    file.write(\"\\n\".join(fileContents))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "# saving file contents list as a .txt file\n",
    "with open(\"SARS_REF_smut_list_amino.txt\", mode=\"w\") as file:\n",
    "    file.write(\"\\n\".join(aminoFileContents))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "aminoFileContents"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
